chatcli_gpt/plugins.py

```diff
@@ -31,33 +39,60 @@ def extract_blocks(response_text, plugin):
     return matches
 
 
+def exec_bash(code):
+    try:
+        result = subprocess.run(
+            ["/bin/bash", "-c", code], capture_output=True, text=True
+        )
+    except Exception:  # pylint: disable=broad-except
+        print(traceback.format_exc())
+
+    return {
+        "result": result.stdout.strip(),
+        "error": result.stderr.strip(),
+    }
+
+
 def exec_python(code):
-    buffer = io.StringIO()
-
-    with contextlib.redirect_stdout(buffer):
-        global_scope = globals()
-        try:
-            mod = ast.parse(code, mode="exec")
-            if isinstance(mod.body[-1], ast.Expr):
-                last_expr = mod.body.pop()
-                exec(compile(mod, "<ast>", "exec"), global_scope)
-                result = eval(compile(ast.Expression(last_expr.value), "<ast>", "eval"), global_scope)
-                if result is not None:
-                    print(result)
-            else:
-                exec(code, global_scope)
-        except Exception:  # pylint: disable=broad-except
-            print(traceback.format_exc())
-
-    return buffer.getvalue().strip()
+    stdout = io.StringIO()
+    stderr = io.StringIO()
+
+    with contextlib.redirect_stdout(stdout):
+        with contextlib.redirect_stderr(stderr):
+            global_scope = globals()
+            try:
+                mod = ast.parse(code, mode="exec")
+                if isinstance(mod.body[-1], ast.Expr):
+                    last_expr = mod.body.pop()
+                    exec(compile(mod, "<ast>", "exec"), global_scope)
+                    result = eval(
+                        compile(ast.Expression(last_expr.value), "<ast>", "eval"),
+                        global_scope,
+                    )
+                    if result is not None:
+                        print(result)
+                else:
+                    exec(code, global_scope)
+            except Exception:  # pylint: disable=broad-except
+                print(traceback.format_exc())
+
+    return {
+        "result": stdout.getvalue().strip(),
+        "error": stderr.getvalue().strip(),
+    }
```

Comment:
I'm mixed on the this. I think the model will be pretty good at figuring out what's error and what's output, and capturing them separately means we lose the correct interleaving?

Reply:
Are you suggesting combining them into one string?
One challenge would be working with something like `curl`, which it writes the progress continuously to `stderr` but emits the result in `stdout`. 
Having these two streams separated, allow the tool to format them (or ignore stderr) later.

Reply:
The way I see it either we keep them together in one string and keep time ordering but lose separation, or split them and lose time ordering.

Yeah, handling big output / error streams is going to a big challenge. I'd hope that curl won't write a progress bar if the output isn't a tty, but I'm sure there'll be plenty of issues like that.

I'm going to merge this and we can consider different approaches in future PRs.